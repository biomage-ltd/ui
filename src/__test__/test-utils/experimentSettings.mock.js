import initialState, {
  initialPipelineState,
  initialWorkerState,
} from '../../redux/reducers/experimentSettings/initialState';

const testableProcessingConfig = {
  cellSizeDistribution: {
    enabled: true,
    auto: true,
    filterSettings: {
      minCellSize: 10800,
      binStep: 200,
    },
  },
  mitochondrialContent: {
    enabled: true,
    auto: true,
    filterSettings: {
      method: 'absolute_threshold',
      methodSettings: {
        absolute_threshold: {
          maxFraction: 0.1,
          binStep: 200,
        },
      },
    },
  },
  classifier: {
    enabled: true,
    auto: true,
    filterSettings: {
      FDR: 0.1,
    },
  },
  numGenesVsNumUmis: {
    enabled: true,
    auto: true,
    filterSettings: {
      regressionType: 'gam',
      regressionTypeSettings: {
        gam: {
          'p.level': 0.00009,
        },
      },
    },
  },
  doubletScores: {
    enabled: true,
    auto: true,
    filterSettings: {
      probabilityThreshold: 0.2,
      binStep: 0.05,
    },
  },
  dataIntegration: {
    enabled: true,
    dataIntegration: {
      method: 'seuratv4',
      methodSettings: {
        seuratv4: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        fastmnn: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        unisample: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
      },
    },
    dimensionalityReduction: {
      method: 'rpca',
      numPCs: 30,
      excludeGeneCategories: [],
    },
  },
  configureEmbedding: {
    enabled: true,
    embeddingSettings: {
      method: 'umap',
      methodSettings: {
        umap: {
          minimumDistance: 0.1,
          distanceMetric: 'euclidean',
        },
        tsne: {
          perplexity: 30,
          learningRate: 200,
        },
      },
    },
    clusteringSettings: {
      method: 'louvain',
      methodSettings: {
        louvain: {
          resolution: 0.5,
        },
      },
    },
  },
};

export default {
  ...initialState,
  processing: {
    ...initialState.processing,
    ...testableProcessingConfig,
  },
};

export { initialPipelineState, initialWorkerState };
