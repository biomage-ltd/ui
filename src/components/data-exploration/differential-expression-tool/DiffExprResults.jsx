import React, { useState, useEffect } from 'react';
import {
  useSelector,
  useDispatch,
} from 'react-redux';
import {
  Space, Button, Alert, Tooltip, Typography,
} from 'antd';
import Link from 'next/link';
import { LeftOutlined } from '@ant-design/icons';
import PropTypes from 'prop-types';
import _ from 'lodash';
import GeneTable from '../generic-gene-table/GeneTable';
import { geneTableUpdateReason } from '../../../utils/geneTable/geneTableUpdateReason';
import loadDifferentialExpression from '../../../redux/actions/differentialExpression/loadDifferentialExpression';

const { Text } = Typography;

const DiffExprResults = (props) => {
  const {
    experimentId, onGoBack, width, height,
  } = props;

  const dispatch = useDispatch();
  const loading = useSelector((state) => state.differentialExpression.properties.loading);
  const data = useSelector((state) => state.differentialExpression.properties.data);
  const total = useSelector((state) => state.differentialExpression.properties.total);
  const error = useSelector((state) => state.differentialExpression.properties.error);
  const comparisonGroup = useSelector((state) => state.differentialExpression.comparison.group);
  const comparisonType = useSelector((state) => state.differentialExpression.comparison.type);

  const properties = useSelector((state) => state.cellSets.properties);

  const [dataShown, setDataShown] = useState(data);
  const [exportAlert, setExportAlert] = useState(false);
  const [settingsListed, setSettingsListed] = useState(false);

  const columns = [
    {
      title: 'Avg log2FC',
      key: 'avg_log2FC',
      sorter: true,
    },
    {
      title: 'adj p-value',
      key: 'p_val_adj',
      sorter: true,
      render: (score, record) => <Tooltip title={`adj p-value: ${record.p_val_adj}`}>{score}</Tooltip>,
    },
    {
      title: 'Pct 1',
      key: 'pct_1',
      sorter: true,
      showSorterTooltip: {
        title: 'The percentage of cells where the feature is detected in the first group',
      },
    },
    {
      title: 'Pct 2',
      key: 'pct_2',
      sorter: true,
      showSorterTooltip: {
        title: 'The percentage of cells where the feature is detected in the second group',
      },
    },
  ];

  // When data changes, update rows.
  useEffect(() => {
    if (data && properties) {
      setDataShown(data);
    }
  }, [data, properties]);

  const isTableLoading = () => data.length === 0 || loading;

  const onUpdate = (newState, reason) => {
    // We handle `loading` and `loaded` in the HOC, no need to react to these.
    if (reason === geneTableUpdateReason.loaded || reason === geneTableUpdateReason.loading) {
      return;
    }

    dispatch(
      loadDifferentialExpression(
        experimentId,
        comparisonGroup[comparisonType],
        comparisonType,
        newState,
      ),
    );
  };

  const optionName = (word) => {
    const [rootGroup, clusterName] = word.split('/');

    let printString = '';

    if (!clusterName) {
      printString = _.capitalize(rootGroup);
    } else if (clusterName === 'rest') {
      printString = `Rest of ${properties[rootGroup].name}`;
    } else {
      printString = properties[clusterName]?.name || _.capitalize(clusterName);
    }

    return <Text strong>{printString}</Text>;
  };

  const { basis, cellSet, compareWith } = comparisonGroup[comparisonType];

  const renderExportAlert = () => {
    if (!exportAlert) return null;
    return (
      <Alert
        message={(
          <span>
            Exporting to CSV is not currently available here. Use the&nbsp;
            <Link
              target='_blank'
              as={`/experiments/${experimentId}/plots-and-tables/volcano`}
              href='/experiments/[experimentId]/plots-and-tables/volcano'
              passHref
            >
              {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
              <a target='_blank'>volcano plot</a>
            </Link>
            &nbsp;in Plots and Tables to export results (opens in new tab).
          </span>
        )}
        type='info'
        closable
        showIcon
        afterClose={() => {
          setExportAlert(false);
        }}
      />
    );
  };

  return (
    <Space direction='vertical' style={{ width: '100%' }}>

      {/* This is needed so changes to the export alert don't cause the table to re-render. */}
      <Space direction='vertical' style={{ width: '100%' }}>
        <Button size='small' onClick={onGoBack}>
          <span>
            <LeftOutlined />
            Go back
          </span>
        </Button>
        {renderExportAlert()}
        <Button id='settingsButton' onClick={() => setSettingsListed(!settingsListed)}>
          {settingsListed ? 'Hide' : 'Show'}
          {' '}
          settings
        </Button>
      </Space>
      {settingsListed
        ? (
          <div id='settingsText'>
            {optionName(cellSet)}
            {' '}
            vs.
            {' '}
            {optionName(compareWith)}
            {' '}
            in
            {' '}
            {optionName(basis)}
          </div>
        ) : <div />}
      <GeneTable
        experimentId={experimentId}
        initialTableState={{
          sorter: {
            field: 'abszscore',
            columnKey: 'abszscore',
            order: 'descend',
          },
        }}
        onUpdate={onUpdate}
        columns={columns}
        loading={isTableLoading()}
        onExportCSV={() => { setExportAlert(true); }}
        error={error}
        width={width}
        height={height - 70 - (exportAlert ? 70 : 0) - (settingsListed ? 70 : 0)}
        data={dataShown}
        total={total}
      />
    </Space>
  );
};

DiffExprResults.defaultProps = {};

DiffExprResults.propTypes = {
  experimentId: PropTypes.string.isRequired,
  onGoBack: PropTypes.func.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default DiffExprResults;
