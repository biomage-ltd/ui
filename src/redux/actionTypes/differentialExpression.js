const DIFF_EXPR = 'diffExpr';

/**
 * Turns on the loading condition for differential expression.
 * Components that have the particular properties requested in their dependencies
 * should set a loading state at this event.
 */
const DIFF_EXPR_LOADING = `${DIFF_EXPR}/loading`;

/**
 * Sets the state of the differential expression data to be successfully loaded.
 * This is meant to be used for paginated data, where the data is
 * lazy-loaded by an API call or equivalent.
 */
const DIFF_EXPR_LOADED = `${DIFF_EXPR}/loaded`;

const DIFF_EXPR_ERROR = `${DIFF_EXPR}/error`;

/**
 * Sets the state for diffential expression settings
 *
 */
const DIFF_EXPR_COMPARISON_TYPE_SET = `${DIFF_EXPR}/comparisonType`;
const DIFF_EXPR_COMPARISON_GROUP_SET = `${DIFF_EXPR}/comparisonGroup`;

export {
  DIFF_EXPR_LOADING, DIFF_EXPR_LOADED, DIFF_EXPR_ERROR,
  DIFF_EXPR_COMPARISON_TYPE_SET, DIFF_EXPR_COMPARISON_GROUP_SET,
};
