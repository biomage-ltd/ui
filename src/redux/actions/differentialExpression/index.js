import loadDifferentialGeneExpression from './loadDifferentialExpression';
import setComparisonType from './setComparisonType';
import setComparisonGroup from './setComparisonGroup';

export {
  loadDifferentialGeneExpression,
  setComparisonType,
  setComparisonGroup,
};
