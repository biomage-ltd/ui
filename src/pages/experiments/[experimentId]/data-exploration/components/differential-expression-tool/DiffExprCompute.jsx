import React, { useState, useEffect } from 'react';
import {
  useSelector, useDispatch,
} from 'react-redux';

import {
  Button, Form, Select, Typography, Radio,
} from 'antd';

import PropTypes from 'prop-types';
import _ from 'lodash';
import { loadCellSets } from '../../../../../../redux/actions/cellSets';
import { setComparisonGroup, setComparisonType } from '../../../../../../redux/actions/differentialExpression';

import composeTree from '../../../../../../utils/composeTree';

const { Text } = Typography;
const { Option, OptGroup } = Select;

const ComparisonType = Object.freeze({ between: 'between', within: 'within' });
const getGroupName = (name) => ( name?.split('/')[0] || null )
const getCellSetName = (name) => ( name?.split('/')[1] || name )

const DiffExprCompute = (props) => {
  const {
    experimentId, onCompute,
  } = props;

  const dispatch = useDispatch();

  const properties = useSelector((state) => state.cellSets.properties);
  const hierarchy = useSelector((state) => state.cellSets.hierarchy);
  const [isFormValid, setIsFormValid] = useState(false);
  const comparisonGroup = useSelector((state) => state.differentialExpression.comparison.group);
  const selectedComparison = useSelector((state) => state.differentialExpression.comparison.type);


  // Metadata marks whether cells belong to the same sample/group
  // Therefore `between samples/groups` analysis makes no sense if there is no metadata.
  // In the base state, assume there is no metadata. A check for this is done in a useEffect block below.
  const [hasMetadata, setHasMetadata] = useState(false);

  /**
   * Loads cell set on initial render if it does not already exist in the store.
   */
  useEffect(() => {
    dispatch(loadCellSets(experimentId));
  }, []);

  useEffect(() => {
    if (hierarchy.length === 0) return;

    // Make sure we are not rendering metadata-related options if there is no metadata in the data set.
    const numMetadata = Object.values(properties).filter((o) => o.type === 'metadataCategorical').length;

    if (!numMetadata) {
      dispatch(setComparisonType(ComparisonType.within));
    }
    setHasMetadata(numMetadata > 0);


    // If any selected option is deleted, set the option to null
    Object.keys(comparisonGroup).forEach((type) => {
      const deleteKeys = {};

      Object.entries(comparisonGroup[type]).forEach(([comparisonKey, selectedCell]) => {
        selectedCell = getCellSetName(selectedCell)
        if(selectedCell || !properties.hasOwnProperty(selectedCell)) deleteKeys[comparisonKey] = null
      });

      if (Object.keys(deleteKeys).length) {
        dispatch(
          setComparisonGroup({
            type,
            ...comparisonGroup[type],
            ...deleteKeys,
          }),
        );
      }

    });

  }, [hierarchy, properties]);

  const validateForm = () => {

    if (!comparisonGroup[selectedComparison]?.cellSet) {
      setIsFormValid(false);
      return;
    }

    if (!comparisonGroup[selectedComparison]?.compareWith) {
      setIsFormValid(false);
      return;
    }

    if (!comparisonGroup[selectedComparison]?.basis) {
      setIsFormValid(false);
      return;
    }

    setIsFormValid(true);
  };

  // Validate form when the groups selected changes.
  useEffect(() => {
    validateForm();
  }, [comparisonGroup[selectedComparison]]);

  /**
   * Updates the selected clusters.
   * @param {string} cellSet The key of the cell set.
   * @param {string} option The option string (`cellSet` or `compareWith`).
   */
  const onSelectCluster = (cellSet, option) => {
    dispatch(setComparisonGroup({
      ...comparisonGroup[selectedComparison],
      type: selectedComparison,
      [option]:
        cellSet,
    }));
  };

  /**
   * Constructs a form item, a `Select` field with selectable clusters.
   */
  const renderClusterSelectorItem = ({
    title, option, filterType,
  }) => {
    // Dependiung on the cell set type specified, set the default name
    const placeholder = filterType === 'metadataCategorical' ? 'sample/group' : 'cell set';

    // Get all the stuff we are going to show.
    const tree = composeTree(hierarchy, properties, filterType);

    const renderChildren = (rootKey, children) => {
      if (!children || children.length === 0) { return (<></>); }

      // If this is the `compareWith` option, we need to add `the rest` under the group previously selected.
      if (option === 'compareWith' && comparisonGroup[selectedComparison]?.cellSet?.startsWith(`${rootKey}/`)) {
        children.unshift({ key: `rest`, name: `Rest of ${properties[rootKey].name}` });
      }

      const shouldDisable = (key) => {
        // Should always disable something already selected.
        if (Object.values(comparisonGroup[selectedComparison]).includes(key)) {
          return true;
        }

        return false;
      }

      if (comparisonGroup[selectedComparison]) {
        return children.map(({ key, name }) => {
          const uniqueKey = `${rootKey}/${key}`;

          return <Option key={uniqueKey} disabled={shouldDisable(uniqueKey)}>
            {name}
          </Option>
        });
      }
    };

    return (
      <Form.Item label={title}>
        <Select
          placeholder={`Select a ${placeholder}...`}
          style={{ width: 200 }}
          onChange={(cellSet) => onSelectCluster(cellSet, option)}
          value={
            comparisonGroup[selectedComparison][option] ?? null
          }
          size='small'
        >
          {
            option === 'basis' &&
            <Option key='all'>
              All
              </Option>
          }
          {
            option === 'compareWith' &&
            <Option key='background'>
              All other cells
              </Option>
          }
          {
            tree && tree.map(({ key, children }) => (
              <OptGroup label={properties[key]?.name} key={key}>
                {renderChildren(key, [...children])}
              </OptGroup>
            ))
          }
        </Select>
      </Form.Item >
    );
  };

  const radioStyle = {
    display: 'block',
    height: '30px',
    lineHeight: '30px',
  };

  return (
    <Form size='small' layout='vertical'>

      <Radio.Group onChange={(e) => {

        dispatch(setComparisonType(e.target.value));

        // If comparisonType key is not available, add to group
        if(!comparisonGroup[e.target.value]) {
          dispatch(setComparisonGroup({
            type: e.target.value,
            cellSet: null,
            compareWith: null,
            basis: null,
            ...comparisonGroup[e.target.value]
          }))
        }

      }} defaultValue={selectedComparison}>
        <Radio
          style={radioStyle}
          value={ComparisonType.between}
          disabled={!hasMetadata}
        >
          Compare a selected cell set between samples/groups
        </Radio>
        <Radio
          style={radioStyle}
          value={ComparisonType.within}>
          Compare cell sets within a sample/group
        </Radio>
      </Radio.Group>

      {selectedComparison === ComparisonType.between
        ? (
          <>
            {renderClusterSelectorItem({
              title: 'Compare cell set:',
              option: 'basis',
              filterType: 'cellSets',
            })}

            {renderClusterSelectorItem({
              title: 'between sample/group:',
              option: 'cellSet',
              filterType: 'metadataCategorical',
            })}

            {renderClusterSelectorItem({
              title: 'and sample/group:',
              option: 'compareWith',
              filterType: 'metadataCategorical',
            })}
          </>
        )
        : (
          <>
            {renderClusterSelectorItem({
              title: 'Compare cell set:',
              option: 'cellSet',
              filterType: 'cellSets',
            })}

            {renderClusterSelectorItem({
              title: 'and cell set:',
              option: 'compareWith',
              filterType: 'cellSets',
            })}

            {renderClusterSelectorItem({
              title: 'within sample/group:',
              option: 'basis',
              filterType: 'metadataCategorical',
            })}
          </>
        )}

      <p>
        <Text type='secondary'>
          Cite
          {' '}
          <a href='https://github.com/kharchenkolab/conos/blob/master/man/Conos.Rd'>
            Conos$getDifferentialGenes
          </a>
          {' '}
          as appropriate.
        </Text>
      </p>

      <Form.Item>
        <Button
          size='small'
          disabled={!isFormValid}
          onClick={() => onCompute()}
        >
          Compute
        </Button>
      </Form.Item>
    </Form>
  );
};

DiffExprCompute.propTypes = {
  experimentId: PropTypes.string.isRequired,
  onCompute: PropTypes.func.isRequired,
};

export default DiffExprCompute;
