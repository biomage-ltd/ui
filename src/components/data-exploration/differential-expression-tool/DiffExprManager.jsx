import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import DiffExprCompute from './DiffExprCompute';
import DiffExprResults from './DiffExprResults';

const DiffExprView = {
  results: 'results',
  compute: 'compute',
};

const DiffExprManager = (props) => {
  const {
    experimentId, view, width, height,
  } = props;

  const [currentView, setCurrentView] = useState(view);

  const onCompute = () => {
    setCurrentView(DiffExprView.results);
  };

  const onGoBack = () => {
    setCurrentView(DiffExprView.compute);
  };

  if (currentView === DiffExprView.compute) {
    return (
      <DiffExprCompute
        experimentId={experimentId}
        onCompute={onCompute}
      />
    );
  }
  if (currentView === DiffExprView.results) {
    return (
      <DiffExprResults
        onGoBack={onGoBack}
        experimentId={experimentId}
        width={width}
        height={height}
      />
    );
  }
};

DiffExprManager.defaultProps = {
  view: DiffExprView.compute,
};

DiffExprManager.propTypes = {
  experimentId: PropTypes.string.isRequired,
  view: PropTypes.string,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default DiffExprManager;
