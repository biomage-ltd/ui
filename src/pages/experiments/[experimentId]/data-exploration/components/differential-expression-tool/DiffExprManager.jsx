import React, { useState } from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import DiffExprCompute from './DiffExprCompute';
import DiffExprResults from './DiffExprResults';

const DiffExprView = {
  results: 'results',
  compute: 'compute',
};

const DiffExprManager = (props) => {
  const {
    experimentId, view, width, height,
  } = props;

  const [selectedCellSets, setSelectedCellSets] = useState({
    cellSet: null,
    compareWith: null,
    basis: null,
  });

  const [currentView, setCurrentView] = useState(view);

  const onCompute = (newSelectedCellSets) => {
    if (!_.isEqual(newSelectedCellSets, selectedCellSets)) {
      setSelectedCellSets(newSelectedCellSets);
    }

    setCurrentView(DiffExprView.results);
  };

  const onGoBack = () => {
    setCurrentView(DiffExprView.compute);
  };

  if (currentView === DiffExprView.compute) {
    return (
      <DiffExprCompute
        experimentId={experimentId}
        onCompute={onCompute}
        cellSets={selectedCellSets}
      />
    );
  }
  if (currentView === DiffExprView.results) {
    return (
      <DiffExprResults
        onGoBack={onGoBack}
        cellSets={{
          cellSet: selectedCellSets.cellSet.split('/')[1],
          compareWith: selectedCellSets.compareWith.split('/')[1] || selectedCellSets.compareWith,
          basis: selectedCellSets.basis.split('/')[1] || selectedCellSets.basis,
        }}
        experimentId={experimentId}
        width={width}
        height={height}
      />
    );
  }
};

DiffExprManager.defaultProps = {
  view: DiffExprView.compute,
};

DiffExprManager.propTypes = {
  experimentId: PropTypes.string.isRequired,
  view: PropTypes.string,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

export default DiffExprManager;
