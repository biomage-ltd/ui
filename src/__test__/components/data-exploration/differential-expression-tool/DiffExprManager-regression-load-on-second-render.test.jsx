import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import preloadAll from 'jest-next-dynamic';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import waitForActions from 'redux-mock-store-await-actions';
import DiffExprManager from '../../../../components/data-exploration/differential-expression-tool/DiffExprManager';
import DiffExprCompute from '../../../../components/data-exploration/differential-expression-tool/DiffExprCompute';
import DiffExprResults from '../../../../components/data-exploration/differential-expression-tool/DiffExprResults';
import initialState from '../../../../redux/reducers/differentialExpression/initialState';

import { DIFF_EXPR_LOADING, DIFF_EXPR_LOADED } from '../../../../redux/actionTypes/differentialExpression';

jest.mock('localforage');

// ensure isBrowser is `true`
jest.mock('../../../../utils/environment', () => ({
  __esModule: true,
  isBrowser: () => true,
}));

jest.mock('../../../../utils/sendWork', () => ({
  __esModule: true, // this property makes it work
  default: jest.fn(() => new Promise((resolve) => resolve({
    results: [
      {
        body: JSON.stringify({
          rows: [
            {
              p_val: 1.4969461240347763e-12, p_val_adj: 1.647289002209057e-11, avg_log2FC: -1.4274754343649423, gene_names: 'A',
            },
            {
              p_val: 2.4969461240347763e-12, p_val_adj: 2.647289002209057e-11, avg_log2FC: -2.4274754343649423, gene_names: 'B',
            },
            {
              p_val: 3.4969461240347763e-12, p_val_adj: 3.647289002209057e-11, avg_log2FC: -3.4274754343649423, gene_names: 'C',
            },
            {
              p_val: 4.4969461240347763e-12, p_val_adj: 4.647289002209057e-11, avg_log2FC: -4.4274754343649423, gene_names: 'D',
            },
            {
              p_val: 5.4969461240347763e-12, p_val_adj: 5.647289002209057e-11, avg_log2FC: -5.4274754343649423, gene_names: 'E',
            },
          ],
          total: 500,
        }),
      },
    ],
  }))),
}));

const mockStore = configureMockStore([thunk]);

const storeState = {
  cellSets: {
    hierarchy: [],
    properties: {},
  },
  genes: {
    focused: undefined,
  },
  differentialExpression: {
    ...initialState,
    comparison: {
      ...initialState.comparison,
      type: 'between',
      group: {
        ...initialState.comparison.group,
        between: {
          ...initialState.comparison.group.between,
          cellSet: 'condition/condition-treated',
          compareWith: 'condition/condition-control',
          basis: 'louvain/cluster-a;',
        },
      },
    },
  },
};

let store = null;

describe('DiffExprManager regression test -- diff exp would not reload after `go back` was hit and a new cluster selected', () => {
  beforeAll(async () => {
    await preloadAll();
  });

  beforeEach(() => {
    store = mockStore(storeState);

    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        document: '',
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  configure({ adapter: new Adapter() });

  it('on click of compute with changed parameters, DiffExprManager calls the results view and dispatches the appropriate actions', async () => {
    const component = mount(
      <Provider store={store}>
        <DiffExprManager experimentId='1234' view='compute' width={100} height={200} />
      </Provider>,
    );

    const cellSets = { cellSet: 'louvain/cluster-1', compareWith: 'rest', basis: 'all' };
    act(() => {
      component.find(DiffExprCompute).props().onCompute('between', cellSets);
    });
    component.update();

    expect(component.find(DiffExprResults).length).toEqual(1);
    expect(component.find(DiffExprCompute).length).toEqual(0);

    await waitForActions(store, [DIFF_EXPR_LOADING, DIFF_EXPR_LOADED]);

    expect(store.getActions().length).toEqual(2);
    expect(store.getActions()[0]).toMatchSnapshot();
    expect(store.getActions()[1]).toMatchSnapshot();
  });

  it('if we then go back and change the parameters again, the new differential expression data should be loading', async () => {
    const component = mount(
      <Provider store={store}>
        <DiffExprManager experimentId='1234' view='compute' width={100} height={200} />
      </Provider>,
    );

    // Choose a cluster and hit compute.
    let cellSets = { cellSet: 'louvain/cluster-2', compareWith: 'rest', basis: 'all' };
    act(() => {
      component.find(DiffExprCompute).props().onCompute('between', cellSets);
    });
    component.update();

    expect(component.find(DiffExprResults).length).toEqual(1);
    expect(component.find(DiffExprCompute).length).toEqual(0);

    // Ensure load diff exp was called.
    await waitForActions(store, [DIFF_EXPR_LOADING, DIFF_EXPR_LOADED]);

    expect(store.getActions().length).toEqual(2);
    expect(store.getActions()[0]).toMatchSnapshot();
    expect(store.getActions()[1]).toMatchSnapshot();

    // Go back.
    act(() => {
      component.find(DiffExprResults).props().onGoBack();
    });
    component.update();

    // Ensure original is still rendered correctly.
    expect(component.find(DiffExprResults).length).toEqual(0);
    expect(component.find(DiffExprCompute).length).toEqual(1);

    // Choose another cell set.
    cellSets = { cellSet: 'louvain/cluster-3', compareWith: 'rest', basis: 'all' };
    act(() => {
      component.find(DiffExprCompute).props().onCompute('between', cellSets);
    });
    component.update();

    // Ensure load diff exp was called again.
    await waitForActions(store, [DIFF_EXPR_LOADING, DIFF_EXPR_LOADED]);

    expect(store.getActions().length).toEqual(4);
    expect(store.getActions()[2]).toMatchSnapshot();
    expect(store.getActions()[3]).toMatchSnapshot();
  });
});
