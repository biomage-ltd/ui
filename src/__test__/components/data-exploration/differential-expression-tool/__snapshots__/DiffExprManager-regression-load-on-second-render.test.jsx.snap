// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffExprManager regression test -- diff exp would not reload after \`go back\` was hit and a new cluster selected if we then go back and change the parameters again, the new differential expression data should be loading 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`DiffExprManager regression test -- diff exp would not reload after \`go back\` was hit and a new cluster selected if we then go back and change the parameters again, the new differential expression data should be loading 2`] = `
Object {
  "payload": Object {
    "cellSets": Object {
      "basis": "louvain/cluster-a;",
      "cellSet": "condition/condition-treated",
      "compareWith": "condition/condition-control",
    },
    "comparisonType": "between",
    "data": Array [
      Object {
        "gene_names": "A",
        "avg_log2FC": -1.4274754343649423,
        "p_val": 1.4969461240347763e-12,
        "p_val_adj": 1.647289002209057e-11,
      },
      Object {
        "gene_names": "B",
        "avg_log2FC": -2.4274754343649425,
        "p_val": 2.4969461240347763e-12,
        "p_val_adj": 2.647289002209057e-11,
      },
      Object {
        "gene_names": "C",
        "avg_log2FC": -3.4274754343649425,
        "p_val": 3.4969461240347765e-12,
        "p_val_adj": 3.647289002209057e-11,
      },
      Object {
        "gene_names": "D",
        "avg_log2FC": -4.4274754343649425,
        "p_val": 4.496946124034777e-12,
        "p_val_adj": 4.647289002209057e-11,
      },
      Object {
        "gene_names": "E",
        "avg_log2FC": -5.4274754343649425,
        "p_val": 5.4969461240347765e-12,
        "p_val_adj": 5.647289002209057e-11,
      },
    ],
    "experimentId": "1234",
    "total": 500,
  },
  "type": "diffExpr/loaded",
}
`;

exports[`DiffExprManager regression test -- diff exp would not reload after \`go back\` was hit and a new cluster selected if we then go back and change the parameters again, the new differential expression data should be loading 3`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`DiffExprManager regression test -- diff exp would not reload after \`go back\` was hit and a new cluster selected if we then go back and change the parameters again, the new differential expression data should be loading 4`] = `
Object {
  "payload": Object {
    "cellSets": Object {
      "basis": "louvain/cluster-a;",
      "cellSet": "condition/condition-treated",
      "compareWith": "condition/condition-control",
    },
    "comparisonType": "between",
    "data": Array [
      Object {
        "gene_names": "A",
        "avg_log2FC": -1.4274754343649423,
        "p_val": 1.4969461240347763e-12,
        "p_val_adj": 1.647289002209057e-11,
      },
      Object {
        "gene_names": "B",
        "avg_log2FC": -2.4274754343649425,
        "p_val": 2.4969461240347763e-12,
        "p_val_adj": 2.647289002209057e-11,
      },
      Object {
        "gene_names": "C",
        "avg_log2FC": -3.4274754343649425,
        "p_val": 3.4969461240347765e-12,
        "p_val_adj": 3.647289002209057e-11,
      },
      Object {
        "gene_names": "D",
        "avg_log2FC": -4.4274754343649425,
        "p_val": 4.496946124034777e-12,
        "p_val_adj": 4.647289002209057e-11,
      },
      Object {
        "gene_names": "E",
        "avg_log2FC": -5.4274754343649425,
        "p_val": 5.4969461240347765e-12,
        "p_val_adj": 5.647289002209057e-11,
      },
    ],
    "experimentId": "1234",
    "total": 500,
  },
  "type": "diffExpr/loaded",
}
`;

exports[`DiffExprManager regression test -- diff exp would not reload after \`go back\` was hit and a new cluster selected on click of compute with changed parameters, DiffExprManager calls the results view and dispatches the appropriate actions 1`] = `
Object {
  "payload": Object {
    "experimentId": "1234",
  },
  "type": "diffExpr/loading",
}
`;

exports[`DiffExprManager regression test -- diff exp would not reload after \`go back\` was hit and a new cluster selected on click of compute with changed parameters, DiffExprManager calls the results view and dispatches the appropriate actions 2`] = `
Object {
  "payload": Object {
    "cellSets": Object {
      "basis": "louvain/cluster-a;",
      "cellSet": "condition/condition-treated",
      "compareWith": "condition/condition-control",
    },
    "comparisonType": "between",
    "data": Array [
      Object {
        "gene_names": "A",
        "avg_log2FC": -1.4274754343649423,
        "p_val": 1.4969461240347763e-12,
        "p_val_adj": 1.647289002209057e-11,
      },
      Object {
        "gene_names": "B",
        "avg_log2FC": -2.4274754343649425,
        "p_val": 2.4969461240347763e-12,
        "p_val_adj": 2.647289002209057e-11,
      },
      Object {
        "gene_names": "C",
        "avg_log2FC": -3.4274754343649425,
        "p_val": 3.4969461240347765e-12,
        "p_val_adj": 3.647289002209057e-11,
      },
      Object {
        "gene_names": "D",
        "avg_log2FC": -4.4274754343649425,
        "p_val": 4.496946124034777e-12,
        "p_val_adj": 4.647289002209057e-11,
      },
      Object {
        "gene_names": "E",
        "avg_log2FC": -5.4274754343649425,
        "p_val": 5.4969461240347765e-12,
        "p_val_adj": 5.647289002209057e-11,
      },
    ],
    "experimentId": "1234",
    "total": 500,
  },
  "type": "diffExpr/loaded",
}
`;
